google_calendar/google_calendar.go

```diff
@@ -0,0 +1,79 @@
+package googlecalendar
+
+// recieve the Google Calendar Push Notifications
+// https://developers.google.com/calendar/v3/push
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	parseheader "github.com/bnfinet/go-httpheader"
+)
+
+// parse errors
+var (
+	ErrInvalidHTTPMethod = errors.New("invalid HTTP Method")
+	ErrParsingPayload    = errors.New("error parsing payload")
+)
+
+// Event defines a Docker hook event type
+type Event string
+
+// GoogleCalendar hook types
+const (
+	SyncEvent      Event = "sync"
+	ExistsEvent    Event = "exists"
+	NotExistsEvent Event = "not_exists"
+)
+
+// GoogleCalendarPayload a google calendar notice
+// https://developers.google.com/calendar/v3/push
+type GoogleCalendarPayload struct {
+	ChannelID         string    `header:"X-Goog-Channel-ID"`
+	ChannelToken      string    `header:"X-Goog-Channel-Token,omitempty"`
+	ChannelExpiration time.Time `header:"X-Goog-Channel-Expiration,omitempty"`
+	ResourceID        string    `header:"X-Goog-Resource-ID"`
+	ResourceURI       string    `header:"X-Goog-Resource-URI"`
+	ResourceState     string    `header:"X-Goog-Resource-State"`
+	MessageNumber     int       `header:"X-Goog-Message-Number"`
+}
+
+// Webhook instance contains all methods needed to process events
+type Webhook struct {
+	secret string
+}
+
+// New creates and returns a WebHook instance
+func New() (*Webhook, error) {
+	hook := new(Webhook)
+	return hook, nil
+}
+
+// Parse verifies and parses the events specified and returns the payload object or an error
+func (hook Webhook) Parse(r *http.Request, events ...Event) (interface{}, error) {
+	defer func() {
+		_, _ = io.Copy(ioutil.Discard, r.Body)
+		_ = r.Body.Close()
+	}()
+
+	if r.Method != http.MethodPost {
+		return nil, ErrInvalidHTTPMethod
+	}
+
+	headers := r.Header
+	if len(headers) == 0 {
+		return nil, ErrParsingPayload
+	}
+
+	gc := &GoogleCalendarPayload{}
+	err := parseheader.ParseHeader(headers, gc)
```

Comment:
what do you think about using `https://github.com/go-playground/form` to parse the headers instead, the main reason being it will return more detailed errors that include the namespace/field names or manually parsing?

Reply:
I've just tried to use `form` however I receive an error when using `decoder.Decode(gc, headers)` in place of `parseheader.ParseHeader(headers, gc)`...

```
var headers http.Header
cannot use headers (variable of type http.Header) as url.Values value in argument to decoder.Decodecompiler
```
 
... and I'm [not finding a method](https://pkg.go.dev/github.com/go-playground/form?utm_source=gopls) which would parse the Headers hashmap.  If it's a matter of avoiding `go-httpheader` I'm happy to parse the headers in some fashion within this codebase.


(later)


Well that was a bit of a rabbit hole...

After working on a new branch for a time which used a variety of calls such as `r.Header.Values("X-Goog-Channel-ID")[0]` to retrieve the headers I think I'm coming down on the side of either 
* a) using a library that relies on reflection such as https://github.com/bnfinet/go-httpheader
* b) integrating the reflection header interpretation into `google_calendar.go`
* c) integrating the reflection header interpretation into `webhooks` overall
* d) possibly expanding `go-playground/form` to include the header based reflection

The struct+reflection based solution is both more resilient and lends itself to be a bit more future proof IMHO.

@deankarn I'd love to get your opinion on an acceptable course before proceeding further